File: config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java
/**
 *   Copyright (C) 2015 Typesafe Inc. <http://typesafe.com>
 */
package com.typesafe.config.impl;

import java.util.*;

abstract class ConfigNodeComplexValue extends AbstractConfigNodeValue {
    final protected ArrayList<AbstractConfigNode> children;

    ConfigNodeComplexValue(Collection<AbstractConfigNode> children) {
        this.children = new ArrayList(children);
    }

    final public Collection<AbstractConfigNode> children() {
        return children;
    }

    @Override
    protected Collection<Token> tokens() {
        ArrayList<Token> tokens = new ArrayList();
        for (AbstractConfigNode child : children) {
            tokens.addAll(child.tokens());
        }
        return tokens;
    }

    protected ConfigNodeComplexValue indentText(AbstractConfigNode indentation) {
        ArrayList<AbstractConfigNode> childrenCopy = new ArrayList<AbstractConfigNode>(children);
        for (int i = 0; i < childrenCopy.size(); i++) {
            if (childrenCopy.get(i) instanceof ConfigNodeSingleToken &&
                    Tokens.isNewline(((ConfigNodeSingleToken) childrenCopy.get(i)).token())) {
                childrenCopy.add(i + 1, indentation);
                i++;
            } else if (childrenCopy.get(i) instanceof ConfigNodeField) {
                AbstractConfigNode value = ((ConfigNodeField) childrenCopy.get(i)).value();
                if (value instanceof ConfigNodeComplexValue) {
                    childrenCopy.set(i, ((ConfigNodeField) childrenCopy.get(i)).replaceValue(((ConfigNodeComplexValue) value).indentText(indentation)));
                }
            } else if (childrenCopy.get(i) instanceof ConfigNodeComplexValue) {
                childrenCopy.set(i, ((ConfigNodeComplexValue) childrenCopy.get(i)).indentText(indentation));
            }
        }
        if (this instanceof ConfigNodeArray) {
            return new ConfigNodeArray(childrenCopy);
        } else if (this instanceof ConfigNodeObject) {
            return new ConfigNodeObject(childrenCopy);
        } else {
            return new ConfigNodeConcatenation(childrenCopy);
        }
    }
}

File: config/src/main/scala/com/typesafe/config/impl/ConfigNodeComplexValue.scala
/**
 *   Copyright (C) 2015 Typesafe Inc. <http://typesafe.com>
 */
package com.typesafe.config.impl

import java.{ util => ju }

abstract class ConfigNodeComplexValue(
    _children: ju.Collection[AbstractConfigNode]) extends AbstractConfigNodeValue {
    // why create ArrayList from Collection?
    val children = new ju.ArrayList[AbstractConfigNode](_children)

    override def tokens: ju.Collection[Token] = {
        val tokens = new ju.ArrayList[Token]
        import scala.collection.JavaConverters._
        for (child <- children.asScala) {
            tokens.addAll(child.tokens)
        }
        tokens
    }
    private[impl] def indentText(indentation: AbstractConfigNode): ConfigNodeComplexValue = {
        val childrenCopy =
            new ju.ArrayList[AbstractConfigNode](children)
        var i = 0
        while (i < childrenCopy.size) {
            val child = childrenCopy.get(i)
            if (child.isInstanceOf[ConfigNodeSingleToken] && Tokens.isNewline(
                child.asInstanceOf[ConfigNodeSingleToken].token)) {
                childrenCopy.add(i + 1, indentation)
                i += 1
            } else if (child.isInstanceOf[ConfigNodeField]) {
                val value =
                    child.asInstanceOf[ConfigNodeField].value
                if (value.isInstanceOf[ConfigNodeComplexValue])
                    childrenCopy.set(
                        i,
                        child
                            .asInstanceOf[ConfigNodeField]
                            .replaceValue(
                                value
                                    .asInstanceOf[ConfigNodeComplexValue]
                                    .indentText(indentation)))
            } else if (child.isInstanceOf[ConfigNodeComplexValue])
                childrenCopy.set(
                    i,
                    child
                        .asInstanceOf[ConfigNodeComplexValue]
                        .indentText(indentation))

            i += 1
        }
        newNode(childrenCopy)
    }
    // This method will just call into the object's constructor, but it's needed
    // for use in the indentText() method so we can avoid a gross if/else statement
    // checking the type of this
    def newNode(nodes: ju.Collection[AbstractConfigNode]): ConfigNodeComplexValue
}
